### PR TITLE
[bug 872708] Added automatic creation of user/get route in requiring app

### PR DESCRIPTION
**Pull request by sedge**
Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=872708

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -72,16 +69,35 @@ module.exports = function ( rawUrl ) {
...
+      }
+      req.session.webmakerid = user.subdomain;
+      res.json( 200, {
+        status: "okay",
+        user: user
+      });
+    });
+  });
+
+  return { Fogin: Fogin, getUser: loginAPI.getUser , isAdmin: loginAPI.isAdmin };
```
Comment (humphd) [inline index.js]:
``` javascript
return {
  Fogin: Fogin,
  getUser: ...
  ...
};
```

File: index.js [hunk 2]
```diff
@@ -72,16 +69,35 @@ module.exports = function ( rawUrl ) {
...
     }
-  };
+  }; // END LoginAPI
+
+  // Routes declaration
+  app.get( "/user/:userid", function( req, res ) {
+    loginAPI.getUser(req.param( 'userid' ), function( err, user ) {
+      if ( err || !user ) {
+        return res.json( 404, {
+        status: "failed",
```
Comment (humphd) [inline index.js]: Indentation failure on these 2 lines (90-1).

File: index.js [hunk 3]
```diff
@@ -72,16 +69,35 @@ module.exports = function ( rawUrl ) {
...
+        reason: ( err || "user not defined" )
+        });
+      }
+      req.session.webmakerid = user.subdomain;
```
Comment (humphd) [inline index.js]: Can you please confirm with Pomax that this is indeed needed/wanted?

File: README.md [hunk 4]
```diff
@@ -7,13 +7,37 @@ Setup is easy!  Add to your package.json and run `npm install`.
...
 
-`require( "webmaker-loginapi" )( "http://admin:roflcoptor@login.wm.org" )`
+`require( "webmaker-loginapi" )( "http://admin:roflcoptor@login.wm.org", expressApp )`
+
+## API Exposure
+
+By passing the express object to this module, you are automatically adding a proxy to this module's `getUser` method through the following route:
+
+```javascript
+app.get( "/user/:userid", function( req, res ) {
```
Comment (humphd) [inline README.md]: Update this to match the code below (indent issue, etc).

File: index.js [hunk 5]
```diff
@@ -72,16 +69,35 @@ module.exports = function ( rawUrl ) {
...
-  };
+  }; // END LoginAPI
+
+  // Routes declaration
+  app.get( "/user/:userid", function( req, res ) {
+    loginAPI.getUser(req.param( 'userid' ), function( err, user ) {
+      if ( err || !user ) {
+        return res.json( 404, {
+        status: "failed",
+        reason: ( err || "user not defined" )
```
Comment (Pomax) [inline index.js]: if it's a 404, let's go with user not found.

File: README.md (hunk 4 above)
Comment (Pomax) [inline README.md]: this does not need to be the full code. Just the route and what property it sets should be sufficient.

File: index.js (hunk 3 above)
Comment (Pomax) [inline index.js]: countering that: let's decide which universal thing we want to go for. I used `webmakerid` "at the time", but we also use `_id` in other places. Which do we pick for homogenization?

Comment (sedge) [inline index.js]: For posterity, we decided on "username".  We also decided to change the user model to reflect this, substituting `username` for `subdomain`

File: README.md [hunk 6]
```diff
@@ -7,15 +7,39 @@ Setup is easy!  Add to your package.json and run `npm install`.
...
 
-## Usage
+`require( "webmaker-loginapi" )( "http://admin:roflcoptor@login.wm.org", expressApp )`
+
+## API Exposure
+
+By passing the express object to this module, you are automatically adding a route that proxies to this module's `getUser` method.  It also sets a session variable called "username":
+
+```javascript
+app.get( "/user/:userid", function( req, res ) {
```
Comment (Pomax) [inline README.md]: I still don't think we need to show actual code here, just say with the route for `/user/:userid` is added, and that it sets `req.session.username` if a user lookup is performed on it, with a 404 if the user was not found

File: index.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ var request = require( "request" ),
     url = require( "url" );
 
 // Module.exports
-module.exports = function ( rawUrl ) {
+module.exports = function ( rawUrl, app ) {
```
Comment (Pomax) [inline index.js]: practicality nit: the URL is going to be long, the app very short, putting 'app' in front is probably going to be nicer. It'll break backward compat, but it's a fix in a line of code that needs fixing in all the apps anyway.

Comment (sedge) [inline index.js]: Okay - I'll make the change.

**Pull request merged**